web/src/ferrostar-core.ts

```diff
@@ -0,0 +1,222 @@
+
+    // Initialize the trip state
+    const startingLocation = this.locationProvider.lastLocation
+      ? this.locationProvider.lastLocation
+      : {
+          coordinates: route.geometry[0],
+          horizontalAccuracy: 0.0,
+          courseOverGround: null,
+          timestamp: Date.now(),
+          speed: null,
+        };
+
+    this.navStateUpdate(
+      this.navigationController.getInitialState(startingLocation),
+    );
+  }
+
+  private saveRecording() {
+    let recording = this.navigationController?.getRecording(
+      this._navState?.recordingEvents,
+    );
+    const blob = new Blob([recording], { type: "application/json" });
+    const link = document.createElement("a");
+    link.href = URL.createObjectURL(blob);
+    // TODO: Figure out how to generate a unique filename
+    link.download = "recording.json";
+    link.click();
+    URL.revokeObjectURL(link.href);
+  }
+
+  async stopNavigation() {
+    if (this.shouldRecord) this.saveRecording();
+    this.routeAdapter?.free();
+    this.routeAdapter = null;
+    this.navigationController?.free();
+    this.navigationController = null;
+    this.navStateUpdate(null);
+    if (this.locationProvider) this.locationProvider.updateCallback = null;
+    if (this.onNavigationStop) this.onNavigationStop();
+  }
+
+  provideState(tripState: TripState) {
+    // Dispatch event for external listeners
+    this.dispatchEvent(
+      new CustomEvent("tripstate-update", {
+        detail: { tripState },
+        bubbles: true,
+      }),
+    );
+  }
+
+  private navStateUpdate(newState: SerializableNavState | null) {
+    this._navState = newState;
+    this.onTripStateChange?.(newState?.tripState || null);
+    if (!newState) return;
+
+    this.provideState(newState.tripState);
+  }
+
+  private onLocationUpdated() {
+    if (!this.navigationController) {
+      return;
+    }
+
+    // Update the trip state with the new location
+    const newNavState = this.navigationController.updateUserLocation(
+      this.locationProvider.lastLocation,
+      this._navState,
+    );
+    this.navStateUpdate(newNavState);
+
+    // Speak the next instruction if voice guidance is enabled
+    const tripState = this._navState?.tripState;
+    if (
+      this.useVoiceGuidance &&
+      tripState != null &&
+      typeof tripState === "object"
+    ) {
+      if (
+        "Navigating" in tripState &&
+        tripState.Navigating?.spokenInstruction &&
+        tripState.Navigating?.spokenInstruction.utteranceId !==
+          this.lastSpokenUtteranceId
+      ) {
+        this.lastSpokenUtteranceId =
+          tripState.Navigating?.spokenInstruction.utteranceId;
+        window.speechSynthesis.cancel();
+        window.speechSynthesis.speak(
+          new SpeechSynthesisUtterance(
+            tripState.Navigating?.spokenInstruction.text,
+          ),
+        );
+      }
+    }
+  }
+
+  // This component doesn't render anything - it's purely for logic
+  render() {
```

Comment:
Discussed on the call: let's try using `ReactiveElement` instead so we don't need a render method.